File: frontend/src/api/action.ts
import { api } from './client'

export interface Action {
  id: string
  chatbot_id: string
  name: string
  description?: string
  action_type: 'builtin' | 'http' | 'zapier'
  config?: any
  parameters?: any
  enabled: boolean
  created_at: string
  updated_at?: string
}

export interface CreateActionRequest {
  name: string
  description?: string
  action_type: 'builtin' | 'http' | 'zapier'
  config?: any
  parameters?: any
  enabled: boolean
}

export const getActions = async (chatbotId: string) => {
  const { data } = await api.get<{ actions: Action[] }>(`/api/v1/chatbots/${chatbotId}/actions`)
  return data.actions
}

export const getAction = async (chatbotId: string, actionId: string) => {
  const { data } = await api.get<Action>(`/api/v1/chatbots/${chatbotId}/actions/${actionId}`)
  return data
}

export const createAction = async (chatbotId: string, action: CreateActionRequest) => {
  const { data } = await api.post<Action>(`/api/v1/chatbots/${chatbotId}/actions`, action)
  return data
}

export const updateAction = async (
  chatbotId: string,
  actionId: string,
  action: Partial<CreateActionRequest>,
) => {
  const { data } = await api.put<Action>(
    `/api/v1/chatbots/${chatbotId}/actions/${actionId}`,
    action,
  )
  return data
}

export const deleteAction = async (chatbotId: string, actionId: string) => {
  await api.delete(`/api/v1/chatbots/${chatbotId}/actions/${actionId}`)
}

export interface ActionLog {
  id: string
  chatbot_id: string
  action_id: string
  conversation_id?: string
  status: string
  request_payload: any
  response_payload: any
  error_message?: string
  duration_ms: number
  created_at: string
}

export const getActionLogs = async (chatbotId: string, page = 1, limit = 20) => {
  const { data } = await api.get<{ logs: ActionLog[]; page: number; limit: number }>(
    `/api/v1/chatbots/${chatbotId}/actions/logs`,
    {
      params: { page, limit },
    },
  )
  return data
}

File: internal/models/action.go
package models

import (
	"encoding/json"
	"time"
)

type ActionType string

const (
	ActionTypeBuiltin ActionType = "builtin"
	ActionTypeHTTP    ActionType = "http"
	ActionTypeZapier  ActionType = "zapier"
)

type ChatbotAction struct {
	ID          string           `json:"id"`
	ChatbotID   string           `json:"chatbot_id"`
	Name        string           `json:"name"`
	Description *string          `json:"description"`
	ActionType  ActionType       `json:"action_type"`
	Config      *json.RawMessage `json:"config"`
	Parameters  *json.RawMessage `json:"parameters"` // JSON Schema
	ToolName    *string          `json:"tool_name"`  // LLM-generated API-compatible identifier
	Enabled     bool             `json:"enabled"`
	Version     int              `json:"version"` // For optimistic locking
	CreatedAt   time.Time        `json:"created_at"`
	UpdatedAt   *time.Time       `json:"updated_at"`
}

// HTTP Action config
type HTTPActionConfig struct {
	URL        string            `json:"url"`
	Method     string            `json:"method"`
	Headers    map[string]string `json:"headers"`
	AuthType   string            `json:"auth_type"` // none, bearer, api_key
	AuthConfig json.RawMessage   `json:"auth_config"`
}

// Zapier Action config
type ZapierActionConfig struct {
	WebhookURL string `json:"webhook_url"`
}

type ActionExecutionLog struct {
	ID              string           `json:"id"`
	ChatbotID       string           `json:"chatbot_id"`
	ActionID        string           `json:"action_id"`
	ConversationID  *string          `json:"conversation_id"`
	MessageID       *string          `json:"message_id"`
	Status          string           `json:"status"` // "success", "failure"
	RequestPayload  *json.RawMessage `json:"request_payload"`
	ResponsePayload *json.RawMessage `json:"response_payload"`
	ErrorMessage    *string          `json:"error_message,omitempty"`
	DurationMs      int              `json:"duration_ms"`
	CreatedAt       time.Time        `json:"created_at"`
}
